### DISCUSSION TITLE
Persistent user session in Unity?

### DISCUSSION THREAD

**Discussion by sanoi** [wiverson/supabase-unity-template#9] [Q&A]
Hey,

I am using Supabase with Unity now. All works fine but I see time to time user session is expired(token expired) and let user login again which is not an ideal situation now. It doesn't matter if I set token expired time to any number, it sometimes keeps the token sometimes not. I followed the template and same result.

BTW, template itself does not keep the user session at all(my local token expired time set to 60 instead of 3600 for test). Any advise on this? Or at least I want to know if this is normal behavior.

**Comment by wiverson** [comment]
The session timeout is set in the server (it’s part of the JWT). Client
just works with that. I usually set the server to the max value.

IIRC there is some logic in the client code to manage refresh. So if the
server is set to max and the user logs in I think at least roughly once a
week it should auto refresh.

--------------------------------------------------------
https://changenode.com/
***@***.***

**Reply by sanoi** [reply]
My expectation(and how it works from other platform, web for example) is
- JWT client session has lifetime set certain amount of time(3600sec by default)
- Expired client session can be refreshed by using refresh token

on top of this basic behaviour, this is what Supabase is doing
- keep client session to local storage
- auto refresh token before it is expired
- refresh token if it is expired

So if everything is okay, user supposed not to login again once they logged in before unless
- auth info removed from server(refresh token doesn't work)
- client local storage reset(client session(token) removed)

With Unity, I see this refresh token doesn't work properly.

I don't want to set client token timeout to max since it is not safe.
I will test more what is actual causing this but please let me know if anyone implemented Unity+Supabase Auth and how it keeps auth session without ask user login again.

**Comment by wiverson** [comment]
That flow is how it’s supposed to work. If you are seeing something else it
could be a lot of things. There’s a background thread to handle token
refresh, a persistence system to save the session token between launches,
etc. unfortunately there’s a lot of stuff that can break including in how
the Unity config is set up, eg if the packages aren’t set up right json
breaks, or if the wrong setting for code stripping, etc etc etc. it’s all
very fragile, alas.

--------------------------------------------------------
https://changenode.com/
***@***.***

**Reply by sanoi** [reply]
Yeah I agree but I followed all from the video and github, I am assuming 2 things now after I checked I failed refresh token with 'Session expired' exception. 

First one is I am using ScriptableObject to manage all Supabase related logic. This might makes all logics run not in main thread.
Second one is UniTask instead of Unity await/async support. I remember it was fine when I use Unity 6 without Unitask, I realize this issue after I change the version to Unity 2023 LTS + UniTask.

I just moved all codes to Monobehavior and testing if I see the same issue. Thanks again for the comment!

**Comment by wiverson** [comment]
I think you have to use a Monobehavior to kick things off, ScriptableObject
doesn’t participate in event loop. You can put methods in an SO but to
invite you need to kick off from a Monobehavior.

Unitask fixes async/await to work properly/better with the event loop. If
you don’t use that you’ll have to reinvent logic to handle thread behavior.

I wouldn’t be surprised if some of the threading behavior changed between
Unity versions. There were a bunch of broken things with Unity threads,
many of which Unitask fixes. Unity nunit in particular is wildly broken wrt
threads.

fwiw I haven’t updated the template since the Unity pricing fiasco. Godot
uses a version of C# much closer to standard, and for my needs it’s been
much easier to work with. I’m writing this from memory while traveling :)


--------------------------------------------------------
https://changenode.com/
***@***.***

**Reply by sanoi** [reply]
Thank you for your comment even you are traveling now. I moved all code to MonoBehaviour but I still have a problem.
At least, now it is consistently failing after client token expired.

As you can see the log below, Debug Listener says 'Loaded session has expired' then it signed out automatically when I called await _supabase.InitializeAsync();

I tried to manually call RefreshToken but it shows error 'GotrueException: Session expired'

I will create completely new Unity project and will check this again. Thank you.

Signed Out
Signed In
Session: Supabase.Gotrue.Session
Initializing Supabase client...
DebugListener====================
Loaded session has expired
Signed Out
Supabase client initialized successfully
Auto-confirm emails on this server: True

`async void Start()
    {
        Debug.Log("SupabaseManager Init");

        Supabase.SupabaseOptions options = new();
        options.AutoConnectRealtime = true;
        options.AutoRefreshToken = true;


        _supabase = new Supabase.Client(SUPABASE_URL, SUPABASE_KEY, options);
        _supabase.Auth.AddDebugListener(DebugListener!);

        _networkStatus.Client = (Supabase.Gotrue.Client)_supabase.Auth;

        _supabase.Auth.SetPersistence(new UnitySession());
        _supabase.Auth.AddStateChangedListener(UnityAuthListener);
        _supabase.Auth.LoadSession();
        Debug.Log("Session: " + _supabase.Auth.CurrentSession);

        // We check the network status to see if we are online or offline using a request to fetch
        // the server settings from our project. Here's how we build that URL.
        string url = $"{SUPABASE_URL}/auth/v1/settings?apikey={SUPABASE_KEY}";
        try
        {
            // This will get the current network status
            _supabase.Auth.Online = await _networkStatus.StartAsync(url);
        }
        catch (NotSupportedException)
        {
            // Some platforms don't support network status checks, so we just assume we are online
            _supabase.Auth.Online = true;
        }
        catch (Exception e)
        {
            // Something else went wrong, so we assume we are offline
            Debug.Log(e.Message);
            Debug.LogException(e);

            _supabase.Auth.Online = false;
        }
        if (_supabase.Auth.Online)
        {
            // Now we start up the client, which will in turn start up the background thread.
            // This will attempt to refresh the session token, which in turn may send a second
            // user login event to the UnityAuthListener.
            try
            {
                Debug.Log("Initializing Supabase client...");
                await _supabase.InitializeAsync();
                Debug.Log("Supabase client initialized successfully");
            }
            catch (Exception e) when (IsTokenExpiredException(e))
            {
                Debug.Log("Token expired during initialization. Attempting to refresh...");
                Debug.LogException(e);
                if (await TryRefreshToken())
                {
                    // Retry initialization after successful token refresh
                    await _supabase.InitializeAsync();
                }
                else
                {
                    // Token refresh failed, need to re-authenticate
                    //await HandleReAuthentication();
                }
            }
            catch (Exception e)
            {
                Debug.LogError($"Failed to initialize Supabase client: {e.Message}");
                // Handle other initialization errors
                Debug.LogException(e);
            }

            // Here we fetch the server settings and log them to the console
            Supabase.Gotrue.Settings serverConfiguration = (await _supabase.Auth.Settings())!;
            Debug.Log($"Auto-confirm emails on this server: {serverConfiguration.MailerAutoConfirm}");
        }
        _supabaseManager.supabaseStarter = this;
    }

    private void OnApplicationQuit()
    {
        if (_supabase != null)
        {
            _supabase?.Auth.Shutdown();
            _supabase = null;
        }
    }
    private void DebugListener(string message, Exception e)
    {
        Debug.Log("DebugListener====================");
        Debug.Log(message);
        // ReSharper disable once ConditionIsAlwaysTrueOrFalseAccordingToNullableAPIContract
        if (e != null)
            Debug.LogException(e);
    }
    private bool IsTokenExpiredException(Exception e)
    {
        // Check if the exception message indicates an expired token
        // You may need to adjust this based on the exact error message you receive
        Debug.Log("IsTokenExpiredException====================");
        Debug.Log(e);
        return e.Message.Contains("token is expired");
    }
    async Task<bool> TryRefreshSession()
    {
        try
        {
            await _supabase.Auth.RefreshSession();
            Debug.Log("Session refreshed successfully");
            return true;
        }
        catch (Exception e)
        {
            Debug.LogError($"Failed to refresh session: {e.Message}");
            Debug.LogException(e);
            return false;
        }
    }
    async Task<bool> TryRefreshToken()
    {
        try
        {
            await _supabase.Auth.RefreshToken();
            Debug.Log("Token refreshed successfully");
            return true;
        }
        catch (Exception e)
        {
            Debug.LogError($"Failed to refresh token: {e.Message}");
            Debug.LogException(e);
            return false;
        }
    }`

**Reply by sanoi** [reply]
I Just checked github project with Unity 2022.3.34f1
After run the scene, signup one user(checked from local server that the user added), screen says I logged in.
But when I stop the player and play again, it says Token expired immediately from DebugListener while await client.InitializeAsync(); was called. I will test more. :(

**Reply by sanoi** [reply]
It is clear that Supabase C# Unity does not refresh token once it is expired for some reason. 100% reproducible by changing token expired time to 60-90 sec, sign-in, close game(stop playing), wait 90 sec and play again then you need to sign in again.

I solve this by calling SetSession right after LoadSession with the values loaded. It is refreshing token while RefreshToken throws error!

**Reply by sanoi** [reply]
```
_supabase = new Supabase.Client(SUPABASE_URL, SUPABASE_KEY, options);
_supabase.Auth.SetPersistence(new UnitySession());
_supabase.Auth.LoadSession();
if (_supabase.Auth.CurrentSession != null)
{
     await _supabase.Auth.SetSession(_supabase.Auth.CurrentSession.AccessToken, 
     _supabase.Auth.CurrentSession.RefreshToken, true);
}
await _supabase.InitializeAsync();
```

        
        
 BTW, I have no idea what InitializeAsync is doing. Token refreshed runtime even without InitializeAsync excuted.

**Reply by wiverson** [reply]
Just got back from vacation.

FWIW you can find all of the source, docs, etc for everything here:
https://github.com/supabase-community/supabase-csharp

...and in the related sub-projects. If you grab the source you can debug/step into everything.

The auth stuff in particular can be found here:
https://github.com/supabase-community/gotrue-csharp

And a lot of info is in the wiki here:
https://github.com/supabase-community/supabase-csharp/wiki

You can do stuff like remove the GoTrue lib from your Unity project and replace it with the GoTrue directory if you are trying to sort out what is going on.

**Reply by sanoi** [reply]
I have checked the code and I think same problem will happen on Godot or any other c# project using this sdk.

When 
await client.InitializeAsync();
is called, inside of the function, it calls await Auth.RetrieveSessionAsync(); at the beginning and this function returns when current session is expired before refreshing it. Which means once user close the app and token in expired, Supabase c# SDK does not refresh the token so user will be logged out all the time.

```
public async Task<Session?> RetrieveSessionAsync()
{
	// No session, so just return.
	if (CurrentSession == null)
		return null;

	// Check to see if the session has expired. If so go ahead and destroy it.
	if (CurrentSession != null && CurrentSession.Expired())
	{
		_debugNotification?.Log($"Loaded session has expired");
		DestroySession();
		return null;
	}

	// If we aren't online, we can't refresh the token
	if (!Online)
	{
		throw new GotrueException("Only supported when online", Offline);
	}

	// We have a session, and hasn't expired, and we seem to be online. Let's try to refresh it.
	if (Options.AutoRefreshToken && CurrentSession?.RefreshToken != null)
	{
		try
		{
			await RefreshToken();
			return CurrentSession;
		}
		catch (Exception e)
		{
			_debugNotification?.Log($"Failed to refresh token ({e.Message})", e);
			_debugNotification?.Log(JsonConvert.SerializeObject(CurrentSession, Formatting.Indented));
			DestroySession();
			return null;
		}
	}

	return CurrentSession;
}
```
		
So the only way to refresh token if it is expired is calling RefreshToken but this is also returns if token is expired which makes refreshing token impossible.

```		
public async Task RefreshToken(string accessToken, string refreshToken)
{
	if (string.IsNullOrEmpty(accessToken) || string.IsNullOrEmpty(refreshToken))
		throw new GotrueException("No token provided", NoSessionFound);

	var result = await _api.RefreshAccessToken(accessToken, refreshToken);

	if (result == null || string.IsNullOrEmpty(result.AccessToken))
		throw new GotrueException("Could not refresh token from provided session.", NoSessionFound);

	CurrentSession = result;
	NotifyAuthStateChange(TokenRefreshed);
}
```

That's why I called SetSession, which is calling RefreshToken(string accessToken, string refreshToken) which is not exposed outside of the SDK and called from SetSession and SetSession doesn't check if current session is expired or not(with force refresh param).

Is there any other way I can achieve this? Or am I misunderstood something? Otherwise C# example/SDK needs to be fixed.

If you want to repro this, you can set token expire time to around 60 sec then login-> quit app -> wait more than 60 sec -> open the app and check if you are still logged in(which means session refreshed after expired). I am pretty sure this is happening on Godot as well.

**Reply by wiverson** [reply]
Normally RefreshToken is used as part of the background thread refresh, not something that the application calls directly.

For more details about usage and also more details check out https://github.com/supabase-community/gotrue-csharp

Look at the test cases in that repo. You'll also see some options for emitting more debug info to help trace what's going on.

Unfortunately each platform has idiosyncrasies as to where it expects data to be persisted, when/how globals are initialized, and how background threads work. To make this easier there are implementations for different stacks (eg Unity, Godot, etc).

FWIW off the cuff IIRC token expiration and refresh threads do have some relationship, some of that is set by the JWT issuer and some of that is set by the refresh thread configuration. You can configure "always broken" relationships, for example if the JWT is set to expire in 60 seconds but the background thread is set to a minimum refresh time of 1 hour, things will of course break.

IIRC, check out the persistence implementation, the background thread config, network monitor implementation, etc. Bunch of stuff that can go sideways.
